### PR TITLE
Mirror codebase structure in tests

### PR DESCRIPTION
**Pull request by Armavica**
**What is this PR about?**

This PR splits up the test files `test_distribution.py`, `test_moments.py` and `test_random.py`, to start tackling issue #5777.

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Major / Breaking Changes
- ...

## Bugfixes / New features
- ...

## Docs / Maintenance
- Break up `test_distribution.py`, `test_moments.py` and `test_random.py` in more specific files.

### CODE REVIEW

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py:1]: This one seems to have logic that makes more sense in other test files, I will leave specific suggestions sometime this week

Comment (ricardoV94) [reply pymc/tests/distributions/test_distribution.py:1]: test_random and test_moments might be better of split across test_discrete/continous/multivariate/...

Comment (Armavica) [reply pymc/tests/distributions/test_distribution.py:1]: Yes, I was indeed thinking to do random and moments next.

**Comment by ricardoV94** [comment]
> This draft PR is a WIP to tackle issue #5777, posted early to avoid duplication of effort.

To avoid conflicts with the other PRs I would suggest we merge this earlier rather than later.

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_logprob`

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_logprob`

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_continuous`

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_continuous` or `dist_math` (wherever it is implemented)

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_continuous`

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_timeseries`

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_multivariate`

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_multivariate`

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be removed (separate commit) or placed in `test_sampling`?

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_multivariate`

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_discrete`

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: I think we are not using this helper anymore, so we can remove it (separate commit) or move it to `util`

**Comment by Armavica** [comment]
> To avoid conflicts with the other PRs I would suggest we merge this earlier rather than later.

Thank you. I don't mind rebasing, but I see what I can do tonight.

**Comment by ricardoV94** [comment]
I didn't mean you have to rush like that. Just giving it priority over other PRs.

**Comment by Armavica** [comment]
The files `test_distribution.py`, `test_moment.py` and `test_random.py` are more or less done. I am just not sure what to do with the final test that remains in `test_random.py`.

I am planning to work on the others this weekend (`gp`, `ode`, `smc`, `step_methods`, `hmc`, `tuning`, `variational`), but you can merge this one now if you want, it is fine for me either way.

File: pymc/tests/distributions/test_random.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_random.py]: I think we can put this in `test_sampling`?

File: pymc/tests/distributions/test_timeseries.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_timeseries.py]: This is repeated.  Also you can simply put the `test_logp` inside the TestGaussianRandomWalk class, not need for the MatchesScipy class. Just call it `test_logp` or something like that

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: Can we put all the other DensityDist tests in this file under this class?

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

**Comment by ricardoV94** [comment]
> I am planning to work on the others this weekend (gp, ode, smc, hmc, tuning, variational), but you can merge this one now if you want, it is fine for me either way.

Yes let's keep this PR just about the distribution tests

**Comment by ricardoV94** [comment]
Looks nearly done. I left some final suggestions. Once you address them I think we can merge.

**Comment by ricardoV94** [comment]
Can you add a bullet point in the maintenance section on the top PR post?

Also some tests seem to be failing, let me know if you need help with those (haven't peaked at the logs)

**Comment by Armavica** [comment]
Thank you for your review! Yes, the tests should be fixable. The only one that I don't understand is the docs: it fails to build the 404 page for an obscure reason... Do you have an idea about this? And it was the same error for the first version of the PR.

**Comment by ricardoV94** [comment]
I think the docs build has been failing everywhere, you can ignore it

**Comment by Armavica** [comment]
I think it is all set now. We should probably think of rebalancing the github test workers at some point, with many tests having moved around.

**Comment by ricardoV94** [comment]
> We should probably think of rebalancing the github test workers at some point, with many tests having moved around.

Yes, after a few more PRs get open/merged we can get an idea of the average runtime to rebalance.

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This seems to be repeated above

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: There's already a TestDensityDist class above

File: pymc/tests/distributions/test_timeseries.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_timeseries.py]: This test is repeated above and below

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

**Comment by Armavica** [comment]
Argh, my bad, tricky rebasing… Good catches.

**Comment by canyon289** [comment]
@Armavica youre doing an amazing job. Thanks for all the effort

File: pymc/tests/distributions/test_continuous.py
Comment (ricardoV94) [inline pymc/tests/distributions/test_continuous.py]: This one is missing an indentation, which is causing the following tests to be inside it and not run (codecov warning)

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [reply pymc/tests/distributions/test_distribution.py]: Still looks unresolved on my end? Codecov is complaining the test above are not being covered, which agress with them still being duplicated.

File: pymc/tests/distributions/test_distribution.py
Comment (ricardoV94) [reply pymc/tests/distributions/test_distribution.py]: Still duplicated above?

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
Still some duplicated tests in `test_distributions` and an indentation error that is putting two tests inside the previous one. I know how painful rebases can be :D

Otherwise looks great!!!

**Comment by Armavica** [comment]
Ok, I understood my mistake: I think I was repeatedly trying to fix the wrong commit. The rebase went without problem this time so I think that it should be ok. And thank you for mentioning codecov: I understood where to look to spot all of this.

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
Merged! Thanks for this :)  Feel free to break some more eggs.

**Pull request merged**
